core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -974,6 +935,28 @@ private StateChangeListener<TaskStatus> createExchangeSinkInstanceHandleUpdateRe
             };
         }
 
+        private void loadMoreTaskDescriptorsIfNecessary()
+        {
+            if (schedulingQueue.size() - schedulingQueue.getSpeculativeCount() < 100) {
```

Comment:
nit: make 100 configurable

Reply:
Correct me if I am wrong here. If we have arbitrary distribution stage with a replicated source then we do not have really much control over how many "speculative" task descriptors we would produce before we enumarte all the splits for the replicated source.
Only after the split enumaration for replicated source completes the priority for tasks will change from "speculative" to "non-speculative".
Probably a corner case which is not something to bother but I would like to be sure I understand the logic here correctly.

Reply:
Yeah. You are right. I think there's even a more generic problem. Currently we do not account memory for task descriptors that are not complete. It is hard to say how big is this problem in practice. The scheduler should not schedule downstream stages before upstream stages are done (at least done with scheduling tasks). Hence in theory the time it takes to enumerate splits for broadcast tasks shouldn't be long. However yeah, today it is inherently racy and it is possible that broadcast split enumeration could be slow. We may need to further improve this algorithm to have more control over it at some point.